pkg/adsc/adsc.go

```diff
@@ -657,51 +666,150 @@ func (a *ADSC) handleLDS(ll []*listener.Listener) {
+	for _, element := range sortTCPListeners {
+		strSer, serr := ConvertProtoMessageToString(a.tcpListeners[element])
+		if serr != nil {
+			adscLog.Warnf("Error for marshaling [%s]: %v", element, serr)
+			continue
+		}
+		sliceItem := "\"" + element + "\":" + strSer
+		sliceTCPListeners = append(sliceTCPListeners, sliceItem)
 	}
-	err = ioutil.WriteFile(base+"_lds_tcp.json", strResponse, 0o644)
+	jsonResponse := "{" + strings.Join(sliceTCPListeners, ",") + "}"
+	err := ioutil.WriteFile(base+"_lds_tcp.json", []byte(jsonResponse), 0o644)
 	if err != nil {
 		return err
 	}
-	strResponse, err = json.MarshalIndent(a.httpListeners, "  ", "  ")
+
+	// guarrante the persistence order for each element in httpListeners
+	var sortHTTPListeners []string
+	for key := range a.httpListeners {
+		sortHTTPListeners = append(sortHTTPListeners, key)
+	}
+	sort.Strings(sortHTTPListeners)
+	var sliceHTTPListeners []string
+	for _, element := range sortHTTPListeners {
+		strSer, serr := ConvertProtoMessageToString(a.httpListeners[element])
+		if serr != nil {
+			adscLog.Warnf("Error for marshaling [%s]: %v", element, serr)
+			continue
+		}
+		sliceItem := "\"" + element + "\":" + strSer
+		sliceHTTPListeners = append(sliceHTTPListeners, sliceItem)
+	}
+	jsonResponse = "{" + strings.Join(sliceHTTPListeners, ",") + "}"
 	if err != nil {
 		return err
 	}
-	err = ioutil.WriteFile(base+"_lds_http.json", strResponse, 0o644)
+	err = ioutil.WriteFile(base+"_lds_http.json", []byte(jsonResponse), 0o644)
 	if err != nil {
 		return err
 	}
-	strResponse, err = json.MarshalIndent(a.routes, "  ", "  ")
+
+	// guarrante the persistence order for each element in routes
+	var sortRoutes []string
+	for key := range a.routes {
+		sortRoutes = append(sortRoutes, key)
+	}
+	sort.Strings(sortRoutes)
+	var sliceRoutes []string
+	for _, element := range sortRoutes {
+		strSer, serr := ConvertProtoMessageToString(a.routes[element])
+		if serr != nil {
+			adscLog.Warnf("Error for marshaling [%s]: %v", element, serr)
+			continue
+		}
+		sliceItem := "\"" + element + "\":" + strSer
+		sliceRoutes = append(sliceRoutes, sliceItem)
+	}
+	jsonResponse = "{" + strings.Join(sliceRoutes, ",") + "}"
 	if err != nil {
 		return err
 	}
-	err = ioutil.WriteFile(base+"_rds.json", strResponse, 0o644)
+	err = ioutil.WriteFile(base+"_rds.json", []byte(jsonResponse), 0o644)
 	if err != nil {
 		return err
 	}
-	strResponse, err = json.MarshalIndent(a.edsClusters, "  ", "  ")
+
+	// guarrante the persistence order for each element in edsClusters
+	var sortEdsClusters []string
+	for key := range a.edsClusters {
+		sortEdsClusters = append(sortEdsClusters, key)
+	}
+	sort.Strings(sortEdsClusters)
+	var sliceEdsClusters []string
+	for _, element := range sortEdsClusters {
+		strSer, serr := ConvertProtoMessageToString(a.edsClusters[element])
+		if serr != nil {
+			adscLog.Warnf("Error for marshaling [%s]: %v", element, serr)
+			continue
+		}
+		sliceItem := "\"" + element + "\":" + strSer
```

Comment:
Would strongly prefer to not have hand crafted JSON. I would suggest following the pattern here: https://github.com/istio/istio/blob/81a739548db1218563822c8fe31101d5108f4658/pilot/pkg/xds/debug.go#L909

Then you can do json.marshal([]jsonMarshalProto) and it will handle it for you

Reply:
okay, this make the code more clear, I define a new type for marshaling,  it's done!